pkg/oomstore/cache.go

```diff
@@ -0,0 +1,173 @@
+package oomstore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/oom-ai/oomstore/internal/database/dbutil"
+	"github.com/oom-ai/oomstore/internal/database/metadata"
+	"github.com/oom-ai/oomstore/internal/database/offline"
+	"github.com/oom-ai/oomstore/pkg/errdefs"
+	"github.com/oom-ai/oomstore/pkg/oomstore/types"
+)
+
+const (
+	Day       = 24 * time.Hour
+	Capacity  = 1000
+	Period    = 5 * time.Minute
+	MinPeriod = 2 * time.Minute
+)
+
+type BufferGroup struct {
+	groupID  int
+	records  []types.StreamRecord
+	lastPush time.Time
+	mu       sync.Mutex
+}
+
+type Cache struct {
+	capacity  int
+	minPeriod time.Duration
+	mu        sync.Mutex
+	quit      chan bool
+
+	ch     chan types.StreamRecord // buffer channel
+	buffer map[int]*BufferGroup    // group_id -> buffered rows
+	ticker *time.Ticker            // group_id -> ticker
+}
+
+func (s *OomStore) InitCache(ctx context.Context) {
+	cache := &Cache{
+		// TODO: make Capacity, Period, MinPeriod configurable
+		capacity:  Capacity,
+		minPeriod: MinPeriod,
+		quit:      make(chan bool),
+
+		ch:     make(chan types.StreamRecord),
+		buffer: make(map[int]*BufferGroup),
+		ticker: time.NewTicker(Period),
+	}
+	s.cache = cache
+
+	go func() {
+		for {
+			select {
+			case <-cache.quit:
+				return
+			case <-cache.ticker.C:
+				for groupID, b := range cache.buffer {
+					if len(b.records) > 0 && time.Since(b.lastPush) > cache.minPeriod {
+						if err := cache.send(ctx, s, groupID); err != nil {
+							fmt.Printf("%+v", err)
```

Comment:
The following line can be used to print error to `stderr` before we have logging framework:
```go
fmt.Fprintf(os.Stderr, "Error pushing to offline store: %+v", err)
```